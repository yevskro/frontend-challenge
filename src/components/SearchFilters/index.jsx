import React from 'react';
import styled from 'styled-components';

function SearchFilters({ children, flexEnd }) {
  return (
    <StyledSearchFilters flexEnd={flexEnd} role="listbox">
      {children}
    </StyledSearchFilters>
  );
}

const StyledSearchFilters = styled.ul`
  display: flex;
  flex-wrap: wrap;
  width: 100%;
  justify-content: ${(props) => (props.flexEnd ? 'flex-end' : 'flex-start')};
  @media (max-width: 415px) {
    justify-content: flex-start;
  }
`;

export default SearchFilters;
