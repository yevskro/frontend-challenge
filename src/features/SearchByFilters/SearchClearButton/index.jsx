import React from 'react';
import styled from 'styled-components';

function SearchClearButton({ onClick }) {
  return <StyledSearchButton onClick={onClick}>Clear</StyledSearchButton>;
}

const StyledSearchButton = styled.button`
  text-align: center;
  font-size: ${({ theme }) => theme.fontSize.medium};
  color: ${({ theme }) => theme.color.neutral.darkGrayishCyan};

  &:hover {
    color: ${({ theme }) => theme.color.primary.desaturtedDarkCyan};
    text-decoration: underline;
    cursor: pointer;
  }
`;

export default SearchClearButton;
