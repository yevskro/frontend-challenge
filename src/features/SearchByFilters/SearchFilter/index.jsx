import React from 'react';
import styled from 'styled-components';

function SearchFilter({
  children,
  filter,
  remove,
  onRemoveFilter,
  onAddFilter,
}) {
  /* a filter that is removeable should only handle an onClick
  on the removing icon.
    a filter that is addable shall not have the remove prop or
    the prop would be set to false. this filter can be clickable
    by the text content
  */
  return (
    <StyledSearchFilter>
      <Text
        onClick={remove ? () => {} : () => onAddFilter(filter)}
        remove={remove}
      >
        {children}
      </Text>
      {remove ? (
        <Remove onClick={() => onRemoveFilter(filter)}>
          <img src="/assets/images/icon-remove.svg" alt="remove icon" />
        </Remove>
      ) : (
        ''
      )}
    </StyledSearchFilter>
  );
}

const StyledSearchFilter = styled.li`
  display: flex;
  justify-content: center;
  align-items: center;
  height: 32px;
  margin: 10px 16px 10px 0px;
  border-radius: 5px;
  overflow: hidden; /* hides the squared border on hover of the span child */
  background-color: ${({ theme }) =>
    theme.color.neutral.lightGrayishCyanFilterTablets};

  @media (min-width: 415px) {
    margin: 10px 0px 10px 16px;
  }
`;

const Text = styled.span`
  display: flex;
  align-items: center;
  height: 100%;
  padding: 0px 9px 0px 8px;
  color: ${({ theme }) => theme.color.primary.desaturtedDarkCyan};
  font-size: ${({ theme }) => theme.fontSize.small};
  font-weight: ${({ theme }) => theme.fontWeight.bold};

  &:hover {
    /* only apply colors if this is a not removable filter */
    color: ${({ remove, theme }) =>
      remove ? theme.color.primary.desaturtedDarkCyan : 'white'};
    background-color: ${({ remove, theme }) =>
      remove
        ? theme.color.neutral.lightGrayishCyanFilterTablets
        : theme.color.primary.desaturtedDarkCyan};
    cursor: ${({ remove }) => (remove ? 'auto' : 'pointer')};
  }
`;

const Remove = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 32px;
  height: 32px;
  border-radius: 0px 5px 5px 0px;
  background-color: ${({ theme }) => theme.color.primary.desaturtedDarkCyan};

  &:hover {
    background-color: ${({ theme }) => theme.color.neutral.veryDarkGrayishCyan};
    cursor: pointer;
  }
`;

export default SearchFilter;
