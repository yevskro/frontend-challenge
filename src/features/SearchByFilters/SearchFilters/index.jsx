import React from 'react';
import styled from 'styled-components';

/* search filters has the ability to flex its children either to flex-end
or flex-start */
function SearchFilters({ children, flexEnd }) {
  return (
    <StyledSearchFilters flexEnd={flexEnd}>{children}</StyledSearchFilters>
  );
}

const StyledSearchFilters = styled.ul`
  display: flex;
  flex-wrap: wrap;
  width: 100%;
  justify-content: ${(props) => (props.flexEnd ? 'flex-end' : 'flex-start')};
  @media (max-width: 415px) {
    justify-content: flex-start;
  }
`;

export default SearchFilters;
