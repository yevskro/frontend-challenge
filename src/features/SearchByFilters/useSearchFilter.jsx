import { useState } from 'react';

function useSearchFilter(_data = [], _filters = []) {
  /* simple search filter one level deep */
  const [filters, setFilters] = useState(_filters);
  const [data] = useState(_data);

  function filterExists(filter, aFilters = filters, deleteIfExists = false) {
    const filterKey =
      typeof filter === 'object' ? Object.keys(filter)[0] : undefined;
    const filterValue = filterKey ? filter[filterKey] : filter;

    for (let idx = 0; idx < aFilters.length; idx += 1) {
      if (filterKey) {
        if (
          typeof filters[idx] === 'object' &&
          filters[idx][filterKey] &&
          filters[idx][filterKey] === filterValue
        ) {
          if (deleteIfExists) {
            aFilters.splice(idx, 1);
          }
          return true;
        }
      } else if (
        typeof filters[idx] !== 'object' &&
        filters[idx] === filterValue
      ) {
        if (deleteIfExists) {
          aFilters.splice(idx, 1);
        }
        return true;
      }
    }

    return false;
  }

  function addFilter(filter) {
    if (!filterExists(filter)) {
      const newFilters = [...filters];
      newFilters.push(filter);
      setFilters(newFilters);
      console.log({ newFilters });
    }
  }

  function removeFilter(filter) {
    const newFilters = [...filters];
    filterExists(filter, newFilters, true);
    setFilters(newFilters);
    console.log(newFilters);
  }

  function getFilters() {
    return filters;
  }

  function getFilteredData() {
    const newData = [];
    let filtered = false;
    for (let dIdx = 0; dIdx < data.length; dIdx += 1) {
      for (let fIdx = 0; fIdx < filters.length; fIdx += 1) {
        if (
          typeof data[dIdx] === 'object' &&
          typeof filters[fIdx] === 'object'
        ) {
          const filtersKey = Object.keys(filters[fIdx])[0];
          const filterValue = filters[fIdx][filtersKey];
          const dataValue = data[dIdx][filtersKey];
          if (dataValue) {
            if (Array.isArray(dataValue) && dataValue.includes(filterValue))
              filtered = true;
          }
          if (
            data[dIdx][filtersKey] &&
            data[dIdx][filtersKey] === filters[fIdx][filtersKey]
          )
            filtered = true;
        }
      }
      if (!filtered) {
        newData.push(data[dIdx]);
      } else filtered = false;
    }
    console.log(newData);
  }

  return { addFilter, removeFilter, getFilters, getFilteredData };
}

export default useSearchFilter;
