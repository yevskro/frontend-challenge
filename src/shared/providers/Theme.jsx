import React from 'react';
import { ThemeProvider } from 'styled-components';
/* theme developed based on the style-guide */
const theme = {
  color: {
    primary: {
      desaturtedDarkCyan: 'hsl(180, 29%, 50%)',
    },
    neutral: {
      lightGrayishCyanBackground: 'hsl(180, 52%, 96%)',
      lightGrayishCyanFilterTablets: 'hsl(180, 31%, 95%)',
      darkGrayishCyan: 'hsl(180, 8%, 52%)',
      veryDarkGrayishCyan: 'hsl(180, 14%, 20%)',
    },
    shadow: 'rgba(94, 140, 160, 0.15)',
  },
  fontSize: {
    xxsmall: '0.7rem',
    xsmall: '0.8rem',
    small: '0.9rem',
    medium: '1rem',
    large: '1.2rem',
  },
  fontWeight: {
    thin: '400',
    normal: '500',
    bold: '600',
  },
  fontFamily: '"Spartan", "sans-serif"',
};

const Theme = ({ children }) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

export default Theme;
