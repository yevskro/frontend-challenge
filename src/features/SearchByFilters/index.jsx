import React from 'react';
import styled from 'styled-components';
import SearchClearButton from './SearchClearButton';

function SearchByFilters({ children, visible, onClear }) {
  return (
    <StyledSearchByFilters visible={visible} role="search">
      {children}
      <SearchClearButton onClick={onClear} />
    </StyledSearchByFilters>
  );
}

const StyledSearchByFilters = styled.div`
  display: flex;
  padding: 10px 25px 10px 21px;
  align-items: center;
  border-radius: 5px;
  background-color: #fff;
  justify-content: space-between;
  box-shadow: 0px 10px 10px ${({ theme }) => theme.color.shadow};
  margin-bottom: 56px;
  visibility: ${({ visible }) => (visible ? 'visibile' : 'hidden')};
  min-height: 75px;
`;

export default SearchByFilters;
