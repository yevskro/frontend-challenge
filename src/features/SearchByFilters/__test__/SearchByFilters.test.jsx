import React from 'react';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Theme from '../../../shared/providers/Theme';
import SearchByFilters from '..';
import SearchFilters from '../SearchFilters';
import SearchFilter from '../SearchFilter';

test('renders', () => {
  render(
    <Theme>
      <SearchByFilters visible>
        <SearchFilters>
          <SearchFilter>CSS</SearchFilter>
        </SearchFilters>
      </SearchByFilters>
    </Theme>
  );

  expect(screen.getByText('CSS')).toBeInTheDocument();
});

test('changes visibility', () => {
  render(
    <Theme>
      <SearchByFilters visible={false}>
        <SearchFilters>
          <SearchFilter>CSS</SearchFilter>
        </SearchFilters>
      </SearchByFilters>
    </Theme>
  );

  expect(screen.getByRole('search')).toHaveStyleRule('visibility', 'hidden');
});

test('snapshot', async () => {
  const tree = renderer
    .create(
      <Theme>
        <SearchByFilters visible>
          <SearchFilters>
            <SearchFilter>CSS</SearchFilter>
          </SearchFilters>
        </SearchByFilters>
      </Theme>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
