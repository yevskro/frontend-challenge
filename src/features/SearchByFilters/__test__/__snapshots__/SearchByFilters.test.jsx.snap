// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
.c4 {
  text-align: center;
  font-size: 1rem;
  color: hsl(180,8%,52%);
}

.c4:hover {
  color: hsl(180,29%,50%);
  -webkit-text-decoration: underline;
  text-decoration: underline;
  cursor: pointer;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-height: 65px;
  padding: 5px 26px 5px 21px;
  margin-bottom: 56px;
  border-radius: 5px;
  background-color: #fff;
  box-shadow: 0px 10px 10px rgba(94,140,160,0.15);
  visibility: visible;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  width: 100%;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 32px;
  margin: 10px 16px 10px 0px;
  border-radius: 5px;
  overflow: hidden;
  background-color: hsl(180,31%,95%);
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
  padding: 0px 9px 0px 8px;
  color: hsl(180,29%,50%);
  font-size: 0.9rem;
  font-weight: 600;
}

.c3:hover {
  color: white;
  background-color: hsl(180,29%,50%);
  cursor: pointer;
}

@media (max-width:415px) {
  .c1 {
    -webkit-box-pack: start;
    -webkit-justify-content: flex-start;
    -ms-flex-pack: start;
    justify-content: flex-start;
  }
}

@media (min-width:415px) {
  .c2 {
    margin: 10px 0px 10px 16px;
  }
}

<div
  className="c0"
  role="search"
>
  <ul
    className="c1"
  >
    <li
      className="c2"
    >
      <span
        className="c3"
        onClick={[Function]}
      >
        CSS
      </span>
      
    </li>
  </ul>
  <button
    className="c4"
  >
    Clear
  </button>
</div>
`;
